pkg/registry/apis/provisioning/register.go

```diff
@@ -477,36 +477,68 @@ func (b *APIBuilder) Validate(ctx context.Context, a admission.Attributes, o adm
 		}
 	}
 
+	// Early exit to avoid more expensive checks if we have already found errors
+	if len(list) > 0 {
+		return invalidRepositoryError(a.GetName(), list)
+	}
+
+	// Exit early if we have already found errors
 	targetError := b.verifyAgaintsExistingRepositories(cfg)
 	if targetError != nil {
-		list = append(list, targetError)
+		return invalidRepositoryError(a.GetName(), field.ErrorList{targetError})
 	}
 
 	// For *create* we do a synchronous test... this can be expensive!
-	// it is the same as a full healthcheck, so should not be run on every update
-	if len(list) == 0 && a.GetOperation() == admission.Create {
-		testResults, err := repository.TestRepository(ctx, repo)
+	// For *update* we only do a test if the generation has changed
+	var shouldTest bool
+	switch a.GetOperation() {
+	case admission.Delete, admission.Connect:
+	case admission.Create:
+		shouldTest = true
+	case admission.Update:
+		oldRepo, err := b.asRepository(ctx, a.GetObject())
 		if err != nil {
-			list = append(list, field.Invalid(field.NewPath("spec"),
-				"Repository test failed", "Unable to verify repository: "+err.Error()))
+			return fmt.Errorf("get old repository for update: %w", err)
 		}
+		oldCfg := oldRepo.Config()
 
-		if !testResults.Success {
-			for _, err := range testResults.Errors {
-				list = append(list, field.Invalid(field.NewPath("spec"),
-					"Repository test failed", err))
-			}
-		}
+		shouldTest = cfg.Generation != oldCfg.Generation
```

Comment:
pretty sure this won't work as expected -- the generation increment happens *after* this call so this would be executed on the following request (likely soon.... given how often this method gets invoked!)